packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js

```diff
@@ -307,15 +308,23 @@ const _writeGranuleFiles = async ({
 
     const granule = await granulePgModel.get(knex, { cumulus_id: granuleCumulusId });
 
+    const errorObject = {
+      Error: 'Failed writing files to Postgres.',
+      Cause: error,
+    };
     await granulePgModel.upsert(
       knex,
       {
         ...granule,
         status: 'failed',
-        error: {
-          Error: 'Failed writing files to Postgres.',
-          Cause: error,
-        },
+        error: errorObject,
+      }
+    );
+    await granuleModel.update(
+      { granuleId: granule.granule_id },
```

Comment:
Should this and the PG upsert above it be run in a `knex.transaction` to ensure that both writes happen or neither do?

Reply:
This function already runs inside a transaction, though. See line 433.

Reply:
Ah, there's a bug here then. The way this is written is a bit confusing. First, we are passing `trx` as a parameter to this function, but not using when doing the PG write on line 316, which is bad for connection use and general logic.

I think creating the transaction on line 433 is a bit superfluous though. I would refactor that to

```
    await _writeGranuleFiles({
      knex,
      granuleCumulusId,
      files,
      workflowStatus,
    });
```

Then surround `granulePgModel.upsert` and `granuleModel.update` in `knex.transaction((trx) => ...` and make sure that `granulePgModel.upsert` uses `trx` instead of `knex`

The whole idea is that we need to use `knex` when writing the files, because we don't want those writes to be transactional so that any writes which can succeed do even if some fail. But I think we would want updating the granule to `failed` in both systems to be transactional

Reply:
🤔 I'm not sure about that.. Say that the update fails in one or the other, that would mean that there would be no database reflection whatsoever of the fact that a write failed, as both records would remain in the `completed` state, correct?

I'll refactor the existing trx because that doesn't make much sense to me either.

Reply:
I guess it's debatable. Every other time we're doing a write to both systems, we're using a transaction because we want to prevent against drift. So we want both writes to succeed or to fail together.

If you leave this implemented as is, then the PG write could succeed, the Dynamo write could fail, and the status would not be the same in both systems. Would that be preferable?

Reply:
In my view, *some* indication that there is a problem is better than none? 🤔

Reply:
Otherwise we are saying "Yes your granule had an error but since we were unable to update both databases we chose not to record this in either database for the sake of preventing drift."

Reply:
I don't necessarily disagree, but it's a departure from how we're doing writes to both systems throughout the rest of the code. I might punt this one to a team vote in the arch meeting

Reply:
So at this point you've already written the granule to two datastores, tried  to update the files  and then when that fails, updating the status of the granules.... you've already written. 

If this somehow fails and we *don't* update the granule with a failed status, I'm wondering if this should throw like it originally did and handle it as a db write failure event and recover it that way.

Reply:
Currently there's a `catch` on both of the writes, so it wouldn't go the DLQ if either write failed @Jkovarik

Reply:
I'm probably entirely missing something, but if the recovery write from the feature brnach throws in the `catch` block, e.g. 

```
  const granule = await granulePgModel.get(knex, { cumulus_id: granuleCumulusId });

    await granulePgModel.upsert(
      knex,
      {
        ...granule,
        status: 'failed',
        error: {
          Error: 'Failed writing files to Postgres.',
          Cause: error,
        },
      }
    );
 ```
 

 it should make it back to the await Promise.allSettled (there's no try/catch that drops the error in `_writeGranule`), which then 

 ```
 const failures = results.filter((result) => result.status === 'rejected');
  if (failures.length > 0) {
    const allFailures = failures.map((failure) => failure.reason);
    const aggregateError = new AggregateError(allFailures);
    log.error('Failed writing some granules to Dynamo', aggregateError);
    throw aggregateError;
  }
  ```

which would eventually wind up back at 

```
 try {
      return await writeRecords({ ...event, cumulusMessage, knex });
    } catch (error) {
      log.fatal(`Writing message failed: ${JSON.stringify(message)}`);
      return sendSQSMessage(process.env.DeadLetterQueue, message);
    }
```

.... resulting in a dead letter?

Reply:
You must not be looking at the most recent code, where there is a `catch` tacked on to the end of that `upsert`:

https://github.com/nasa/cumulus/blob/e2df6c4f11d1f210e74c3e2691f76d3b2e150c48/packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js#L322

Reply:
😓  (ugh, right you are, I was locally reviewing a branch  I'd wrongly assumed had pulled in the feature branch) 

Good clarification, thanks.    I stand by the question, though, sans the `like it originally did`.